File: backend/src/controllers/studySessionController.ts
import { Request, Response } from 'express';
import { z } from 'zod';
import { StudySessionService } from '../services/studySessionService.js';

/**
 * Study session controller handling session management and question flow
 * Implements weighted question selection with confidence-based learning
 */
export class StudySessionController {
  private studySessionService = new StudySessionService();

  /**
   * Start or resume a study session
   */
  startSession = async (req: Request, res: Response): Promise<void> => {
    try {
      const startSessionSchema = z.object({
        questionSetId: z.number(),
        mode: z.enum(['front-to-end', 'shuffle']),
      });

      const validatedData = startSessionSchema.parse(req.body);
      const userId = req.user?.userId;

      if (!userId) {
        res.status(401).json({ error: 'Unauthorized' });
        return;
      }

      const session = await this.studySessionService.startOrResumeSession(userId, validatedData);

      res.json({
        success: true,
        session: {
          id: session.id,
          questionSetId: session.questionSetId,
          mode: session.mode,
          startedAt: session.startedAt,
          isResumed: session.sessionAnswers.length > 0,
        },
      });
    } catch (error) {
      console.error('Error starting study session:', error);
      if (error instanceof z.ZodError) {
        res.status(400).json({ error: 'Invalid request data', details: error.errors });
        return;
      }
      res.status(500).json({ error: 'Failed to start study session' });
    }
  };

  /**
   * Get the next question in the study session
   */
  getNextQuestion = async (req: Request, res: Response): Promise<void> => {
    try {
      const { questionSetId } = req.params;
      const userId = req.user?.userId;

      if (!userId) {
        res.status(401).json({ error: 'Unauthorized' });
        return;
      }

      const questionSetIdNum = parseInt(questionSetId);
      if (isNaN(questionSetIdNum)) {
        res.status(400).json({ error: 'Invalid question set ID' });
        return;
      }

      const result = await this.studySessionService.getNextQuestion(userId, questionSetIdNum);

      res.json({
        success: true,
        ...result,
      });
    } catch (error) {
      console.error('Error getting next question:', error);
      res.status(500).json({ error: 'Failed to get next question' });
    }
  };

  /**
   * Submit an answer with confidence rating
   */
  submitAnswer = async (req: Request, res: Response): Promise<void> => {
    try {
      const submitAnswerSchema = z.object({
        questionId: z.number(),
        confidenceRating: z.number().min(1).max(5),
      });

      const { questionSetId } = req.params;
      const validatedData = submitAnswerSchema.parse(req.body);
      const userId = req.user?.userId;

      if (!userId) {
        res.status(401).json({ error: 'Unauthorized' });
        return;
      }

      const questionSetIdNum = parseInt(questionSetId);
      if (isNaN(questionSetIdNum)) {
        res.status(400).json({ error: 'Invalid question set ID' });
        return;
      }

      await this.studySessionService.submitAnswer(userId, questionSetIdNum, validatedData);

      res.json({
        success: true,
        message: 'Answer submitted successfully',
      });
    } catch (error) {
      console.error('Error submitting answer:', error);
      if (error instanceof z.ZodError) {
        res.status(400).json({ error: 'Invalid request data', details: error.errors });
        return;
      }
      res.status(500).json({ error: 'Failed to submit answer' });
    }
  };

  /**
   * Complete current study session
   */
  completeSession = async (req: Request, res: Response): Promise<void> => {
    try {
      const { questionSetId } = req.params;
      const userId = req.user?.userId;

      if (!userId) {
        res.status(401).json({ error: 'Unauthorized' });
        return;
      }

      const questionSetIdNum = parseInt(questionSetId);
      if (isNaN(questionSetIdNum)) {
        res.status(400).json({ error: 'Invalid question set ID' });
        return;
      }

      await this.studySessionService.completeSession(userId, questionSetIdNum);

      res.json({
        success: true,
        message: 'Study session completed',
      });
    } catch (error) {
      console.error('Error completing study session:', error);
      res.status(500).json({ error: 'Failed to complete study session' });
    }
  };

  /**
   * Restart study session (creates new session)
   */
  restartSession = async (req: Request, res: Response): Promise<void> => {
    try {
      const restartSessionSchema = z.object({
        mode: z.enum(['front-to-end', 'shuffle']),
      });

      const { questionSetId } = req.params;
      const validatedData = restartSessionSchema.parse(req.body);
      const userId = req.user?.userId;

      if (!userId) {
        res.status(401).json({ error: 'Unauthorized' });
        return;
      }

      const questionSetIdNum = parseInt(questionSetId);
      if (isNaN(questionSetIdNum)) {
        res.status(400).json({ error: 'Invalid question set ID' });
        return;
      }

      const session = await this.studySessionService.restartSession(userId, {
        questionSetId: questionSetIdNum,
        mode: validatedData.mode,
      });

      res.json({
        success: true,
        session: {
          id: session.id,
          questionSetId: session.questionSetId,
          mode: session.mode,
          startedAt: session.startedAt,
        },
      });
    } catch (error) {
      console.error('Error restarting study session:', error);
      if (error instanceof z.ZodError) {
        res.status(400).json({ error: 'Invalid request data', details: error.errors });
        return;
      }
      res.status(500).json({ error: 'Failed to restart study session' });
    }
  };

  /**
   * Reset study session - complete deletion and fresh start
   */
  resetSession = async (req: Request, res: Response): Promise<void> => {
    try {
      const resetSessionSchema = z.object({
        mode: z.enum(['front-to-end', 'shuffle']).optional().default('front-to-end'),
      });

      const { questionSetId } = req.params;
      const validatedData = resetSessionSchema.parse(req.body);
      const userId = req.user?.userId;

      if (!userId) {
        res.status(401).json({ error: 'Unauthorized' });
        return;
      }

      const questionSetIdNum = parseInt(questionSetId);
      if (isNaN(questionSetIdNum)) {
        res.status(400).json({ error: 'Invalid questionSetId' });
        return;
      }

      console.log('🔄 StudySessionController: Resetting session for questionSet', questionSetIdNum, 'user', userId);

      const session = await this.studySessionService.resetSession(
        userId,
        questionSetIdNum,
        validatedData.mode
      );

      res.json({
        success: true,
        message: 'Session reset successfully',
        session: {
          id: session.id,
          questionSetId: session.questionSetId,
          mode: session.mode,
          startedAt: session.startedAt,
        },
      });
    } catch (error) {
      console.error('Error resetting study session:', error);
      if (error instanceof z.ZodError) {
        res.status(400).json({ error: 'Invalid request data', details: error.errors });
        return;
      }
      res.status(500).json({ error: 'Failed to reset study session' });
    }
  };

  /**
   * Get current study session status
   */
  getSessionStatus = async (req: Request, res: Response): Promise<void> => {
    try {
      const { questionSetId } = req.params;
      const userId = req.user?.userId;

      if (!userId) {
        res.status(401).json({ error: 'Unauthorized' });
        return;
      }

      const questionSetIdNum = parseInt(questionSetId);
      if (isNaN(questionSetIdNum)) {
        res.status(400).json({ error: 'Invalid question set ID' });
        return;
      }

      // Get next question which includes progress information
      const result = await this.studySessionService.getNextQuestion(userId, questionSetIdNum);

      res.json({
        success: true,
        hasActiveSession: !result.sessionComplete,
        progress: result.progress,
        sessionComplete: result.sessionComplete,
      });
    } catch (error) {
      // If no active session, return appropriate response
      if (error instanceof Error && error.message === 'No active study session found') {
        res.json({
          success: true,
          hasActiveSession: false,
          progress: null,
          sessionComplete: false,
        });
        return;
      }

      console.error('Error getting session status:', error);
      res.status(500).json({ error: 'Failed to get session status' });
    }
  };
}

File: backend-springboot/src/main/java/com/learningadvisor/backend/controller/StudySessionController.java
package com.learningadvisor.backend.controller;

import com.learningadvisor.backend.dto.studysession.*;
import com.learningadvisor.backend.service.StudySessionService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

/**
 * Study session controller handling session management and question flow
 * Implements weighted question selection with confidence-based learning
 */
@RestController
@RequestMapping("/api/study-sessions")
@RequiredArgsConstructor
@Slf4j
public class StudySessionController {

    private final StudySessionService studySessionService;

    /**
     * Extract user ID from security context
     */
    private Long getUserIdFromContext() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication != null && authentication.getPrincipal() instanceof Long) {
            return (Long) authentication.getPrincipal();
        }
        throw new RuntimeException("Unauthorized");
    }

    /**
     * Start or resume a study session
     */
    @PostMapping("/start")
    public ResponseEntity<Map<String, Object>> startSession(
            @Valid @RequestBody CreateStudySessionRequest request) {
        try {
            Long userId = getUserIdFromContext();
            StudySessionResponse session = studySessionService.startOrResumeSession(userId, request);

            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("session", session);

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            log.error("Error starting study session", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to start study session");
            return ResponseEntity.status(500).body(error);
        }
    }

    /**
     * Get session status for a question set
     */
    @GetMapping("/{questionSetId}/status")
    public ResponseEntity<Map<String, Object>> getSessionStatus(
            @PathVariable Long questionSetId) {
        try {
            Long userId = getUserIdFromContext();
            NextQuestionResponse result = studySessionService.getNextQuestion(userId, questionSetId);

            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("hasActiveSession", !result.getSessionComplete());
            response.put("progress", result.getProgress());
            response.put("sessionComplete", result.getSessionComplete());

            return ResponseEntity.ok(response);
        } catch (RuntimeException e) {
            if (e.getMessage().equals("No active study session found")) {
                Map<String, Object> response = new HashMap<>();
                response.put("success", true);
                response.put("hasActiveSession", false);
                response.put("progress", null);
                response.put("sessionComplete", false);
                return ResponseEntity.ok(response);
            }

            log.error("Error getting session status", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to get session status");
            return ResponseEntity.status(500).body(error);
        }
    }

    /**
     * Get next question in session
     */
    @GetMapping("/{questionSetId}/next-question")
    public ResponseEntity<Map<String, Object>> getNextQuestion(
            @PathVariable Long questionSetId) {
        try {
            Long userId = getUserIdFromContext();
            NextQuestionResponse result = studySessionService.getNextQuestion(userId, questionSetId);

            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("question", result.getQuestion());
            response.put("questionNumber", result.getQuestionNumber());
            response.put("previousScore", result.getPreviousScore());
            response.put("sessionComplete", result.getSessionComplete());
            response.put("progress", result.getProgress());

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            log.error("Error getting next question", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to get next question");
            return ResponseEntity.status(500).body(error);
        }
    }

    /**
     * Submit answer with confidence rating
     */
    @PostMapping("/{questionSetId}/submit-answer")
    public ResponseEntity<Map<String, Object>> submitAnswer(
            @PathVariable Long questionSetId,
            @Valid @RequestBody SubmitAnswerRequest request) {
        try {
            Long userId = getUserIdFromContext();
            studySessionService.submitAnswer(userId, questionSetId, request);

            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Answer submitted successfully");

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            log.error("Error submitting answer", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to submit answer");
            return ResponseEntity.status(500).body(error);
        }
    }

    /**
     * Complete current session
     */
    @PostMapping("/{questionSetId}/complete")
    public ResponseEntity<Map<String, Object>> completeSession(
            @PathVariable Long questionSetId) {
        try {
            Long userId = getUserIdFromContext();
            studySessionService.completeSession(userId, questionSetId);

            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Study session completed");

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            log.error("Error completing study session", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to complete study session");
            return ResponseEntity.status(500).body(error);
        }
    }

    /**
     * Restart session with new mode
     */
    @PostMapping("/{questionSetId}/restart")
    public ResponseEntity<Map<String, Object>> restartSession(
            @PathVariable Long questionSetId,
            @Valid @RequestBody CreateStudySessionRequest request) {
        try {
            Long userId = getUserIdFromContext();

            // Override questionSetId from path
            request.setQuestionSetId(questionSetId);

            StudySessionResponse session = studySessionService.restartSession(userId, request);

            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("session", session);

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            log.error("Error restarting study session", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to restart study session");
            return ResponseEntity.status(500).body(error);
        }
    }

    /**
     * Reset session - complete deletion and fresh start
     */
    @PostMapping("/{questionSetId}/reset")
    public ResponseEntity<Map<String, Object>> resetSession(
            @PathVariable Long questionSetId,
            @RequestBody(required = false) Map<String, String> requestBody) {
        try {
            Long userId = getUserIdFromContext();
            String mode = requestBody != null && requestBody.containsKey("mode")
                    ? requestBody.get("mode")
                    : "front-to-end";

            log.info("Reset session for questionSet {} user {}", questionSetId, userId);

            StudySessionResponse session = studySessionService.resetSession(userId, questionSetId, mode);

            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Session reset successfully");
            response.put("session", session);

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            log.error("Error resetting study session", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to reset study session");
            return ResponseEntity.status(500).body(error);
        }
    }

    /**
     * Get questions with selection probabilities for sidebar
     */
    @GetMapping("/{questionSetId}/questions-probabilities")
    public ResponseEntity<Map<String, Object>> getQuestionsWithProbabilities(
            @PathVariable Long questionSetId) {
        try {
            Long userId = getUserIdFromContext();
            QuestionProbabilitiesResponse result = studySessionService.getQuestionsWithProbabilities(
                    userId, questionSetId);

            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("questions", result.getQuestions());
            response.put("totalWeight", result.getTotalWeight());
            response.put("currentQuestionId", result.getCurrentQuestionId());

            return ResponseEntity.ok(response);
        } catch (RuntimeException e) {
            if (e.getMessage().equals("No active session found")) {
                Map<String, Object> error = new HashMap<>();
                error.put("error", "No active session found");
                return ResponseEntity.status(404).body(error);
            }

            log.error("Error getting questions with probabilities", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to get questions with probabilities");
            return ResponseEntity.status(500).body(error);
        }
    }

    /**
     * Select a specific question for study (if eligible)
     */
    @PostMapping("/{questionSetId}/select-question")
    public ResponseEntity<Map<String, Object>> selectQuestion(
            @PathVariable Long questionSetId,
            @Valid @RequestBody SelectQuestionRequest request) {
        try {
            Long userId = getUserIdFromContext();
            NextQuestionResponse result = studySessionService.selectQuestion(
                    userId, questionSetId, request.getQuestionId());

            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("question", result.getQuestion());
            response.put("questionNumber", result.getQuestionNumber());
            response.put("previousScore", result.getPreviousScore());
            response.put("sessionComplete", result.getSessionComplete());
            response.put("progress", result.getProgress());

            return ResponseEntity.ok(response);
        } catch (RuntimeException e) {
            if (e.getMessage().contains("not selectable") || e.getMessage().contains("not found")) {
                Map<String, Object> error = new HashMap<>();
                error.put("error", e.getMessage());
                return ResponseEntity.status(400).body(error);
            }
            if (e.getMessage().equals("No active session found")) {
                Map<String, Object> error = new HashMap<>();
                error.put("error", "No active session found");
                return ResponseEntity.status(404).body(error);
            }

            log.error("Error selecting question", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to select question");
            return ResponseEntity.status(500).body(error);
        }
    }

    /**
     * Get hypothetical probabilities for live updates
     */
    @PostMapping("/{questionSetId}/hypothetical-probabilities")
    public ResponseEntity<Map<String, Object>> getQuestionsWithHypotheticalProbabilities(
            @PathVariable Long questionSetId,
            @Valid @RequestBody HypotheticalProbabilitiesRequest request) {
        try {
            Long userId = getUserIdFromContext();
            QuestionProbabilitiesResponse result = studySessionService.getQuestionsWithHypotheticalProbabilities(
                    userId, questionSetId, request.getQuestionId(), request.getHypotheticalRating());

            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("questions", result.getQuestions());
            response.put("totalWeight", result.getTotalWeight());
            response.put("currentQuestionId", result.getCurrentQuestionId());

            return ResponseEntity.ok(response);
        } catch (RuntimeException e) {
            if (e.getMessage().equals("No active session found")) {
                Map<String, Object> error = new HashMap<>();
                error.put("error", "No active session found");
                return ResponseEntity.status(404).body(error);
            }

            log.error("Error getting hypothetical probabilities", e);
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to get hypothetical probabilities");
            return ResponseEntity.status(500).body(error);
        }
    }
}
